### DISCUSSION TITLE
Where are the Remote ID and Local ID in the server config

### DISCUSSION THREAD

**Discussion by daniel-dlds** [gitbls/pistrong#13] [General]
Hello,

I have a strongswan managed by pistrong that is exposed on the net. Now I want to connect to it from my mac laptop. The VPN has two fields : Remote ID and Local ID.
Where are these values configured in the configuration files ?

Thanks,
Regards

**Comment by gitbls** [comment]
I'm not completely clear what you're asking, but here's my interpretation of it. You defined the remote IDs when you ran MakeMyCA. There are remote IDs for each of android, ios, windows, and linux. You can see them in the `sudo pistrong config` command, item validids. They are also visible in the config file /etc/swanctl/conf.d/pistrong-CAServerConnection.conf

The Local ID is per-user. That is, each user has a specific Local ID. You can see that by `sudo pistrong list username-devname --full` in the field AltNames

Once you create a Cert for your MacOS user, the file /etc/swanctl/pistrong/server-assets/username-devname.zip is created, which contains the certs for that specific client. There is also a readme.txt file in the zip file that contains the Remote ID and Local ID for that particular Cert. Also note that the Local ID for a cert is fabricated from the inputs you provided to MakeMyCA and the usernname and devname for that particular Cert.

Hope the above answers your question. If not, please respond with more specifics of what you're looking for.

**Comment by daniel-dlds** [comment]
Did a test with my laptop connected to 4G and kept an eye on the vpn server log. This is what I am getting :

`Mar 11 18:41:12 vpnserver charon: 02[NET] received packet: from XXX.XXX.XXX.XXX[500] to 192.168.1.100[500] (604 bytes)
Mar 11 18:41:12 vpnserver charon: 02[ENC] parsed IKE_SA_INIT request 0 [ SA KE No N(REDIR_SUP) N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) ]
Mar 11 18:41:12 vpnserver charon: 02[IKE] XXX.XXX.XXX.XXX is initiating an IKE_SA
Mar 11 18:41:12 vpnserver charon: 02[CFG] selected proposal: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048
Mar 11 18:41:12 vpnserver charon: 02[IKE] local host is behind NAT, sending keep alives
Mar 11 18:41:12 vpnserver charon: 02[IKE] remote host is behind NAT
Mar 11 18:41:12 vpnserver charon: 02[IKE] sending cert request for "C=US, O=vpnserver-strongSwan, CN=strongSwan vpnserver Root CA"
Mar 11 18:41:12 vpnserver charon: 02[ENC] generating IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) CERTREQ N(FRAG_SUP) N(CHDLESS_SUP) N(MULT_AUTH) ]
Mar 11 18:41:12 vpnserver charon: 02[NET] sending packet: from YYY.YYY.YYY.YYY[500] to XXX.XXX.XXX.XXX[500] (481 bytes)
Mar 11 18:41:32 vpnserver charon: 12[IKE] sending keep alive to XXX.XXX.XXX.XXX[500]
Mar 11 18:41:42 vpnserver charon: 15[JOB] deleting half open IKE_SA with XXX.XXX.XXX.XXX after timeout
`
where XXX.XXX.XXX.XXX is the public IP address of the client and YYY.YYY.YYY.YYY is the internal ip address of the VPN server.
What can I do to get around this ?

**Comment by gitbls** [comment]
Unfortunately, your log segment doesn't tell me anything useful, other than it appears your client does not have any Certificates installed on it.

Here's what I need from you in order to assist you:

- What is your client OS?
- Did you accurately and completely follow the directions for that OS to install the  Client certs on that system, as documented [here](https://github.com/gitbls/pistrong/blob/master/CertInstall.md)?
- Please do the following:
    - Download pscollect from this github to your VPN Server
    - `sudo chmod 755 pscollect`
    - run it with `sudo pscollect`
    - copy the output to here in such a way that the line breaks are retained;. (you did not do this in the log output above).
    - If you'd prefer to not post the output here, you can email it to gitbls@outlook.com, but I will still need the line breaks maintained.

Based on the output you provide to the above, I may also ask you to provide some output from your Client OS to ascertain whether there is a problem with the Cert installs on the Client.

**Comment by gitbls** [comment]
Hello @daniel-dlds, waiting to hear back from you on this.

**Comment by daniel-dlds** [comment]
Hello Benn, just got back to this, sorry for the delay.
The client is a mac.
I regenerated the CA certificates with makeMyCA because my domain name was not correct.
I ran ps collect and it output the following to the terminal :

*** Gathering diagnostic information for pistrong/strongSwan ***
...Environmental details
...apt-cache policy strongswan
...systemctl status strongswan
...pistrong config [minus smtp credentials]
...pistrong showca
...pistrong list --full
...copy /etc/swanctl/pistrong/makeCA.log
...copy /etc/swanctl/conf.d/pistrong-CAServerConnection.conf
...ls -lR /etc/swanctl
...ip configuration information
Error: ipv4: FIB table does not exist.
Dump terminated
...iptables filter table
...iptables nat table
...journalctl for charon and swanctl
Done.
Please provide /tmp/pistrong-diagnostics to the pistrong maintainer as requested

Last time there were no line breaks because apparently the code button does not put them, so now it goes like plain text.

[pistrong-diagnostics.zip](https://github.com/gitbls/pistrong/files/11105592/pistrong-diagnostics.zip)

**Comment by gitbls** [comment]
Hi, thanks for posting the diagnostics. I odn't see anything terribly wrong there. However, there seem to be a veritable s**-load of VPN connect requests, and I can't tell which one(s), if any, are from your Mac. 

Can you please try the following:

- `sudo systemctl stop strongswan`
- Get your Mac ready to click on the VPN 'connect' button
- `sudo logger "Ready to do VPN connect from Mac"
- `sudo systemctl start strongswan`
- Press connect on the Mac "connect VPN button
- Wait 10 seconds
- `sudo logger "Mac VPN connect attempt done"
- What is the exact error text reported on the Mac when the VPN connection fails?

I suspect you'll find NO connection attempt in the system log on your Pi (`sudo journalctl -b`). I'm not Mac-fluent, but your Mac VPN connection seems to be wrong in some way. Is your Mac on the same LAN as the Pi, or elsewhere on the internet? If it's on the LAN, is the DNS name vpnserver.zapto.org defined on your LAN such that the Mac can find it?

Did you create a Cert for the mac using `sudo pistrong add dlsa ...` and then install it on the Mac? The client certs would show up in the directory /etc/swanctl/pistrong/server-assets as a zip file, and there should also be a .p12 file in the directory /etc/swanctl/p12, and a corresponding .pem file in /etc/swanctl/private. These would show up after you do the `pistrong add` command. I'm not seeing any of that. 

I note that the DNS name `vpnserver.zapto.org` is not defined on the internet, so if your Mac is outside your LAN, it will fail for that reason as well.

There's no magic here. The DNS plumbing must be correct for things to work, the Mac VPN configuration needs to be correct. and the Certs need to be correctly created and installed on your Mac.

I also note that your VPN server has both hardwired ethernet and WiFi connected. Normally RasPiOS will use ethernet in a case like this,  Why do you have WiFi enabled on this system? All it's doing is creating a lot of log spew and making things more confusing. 

What is the result if you turn off your WiFi, reboot (with WiFi disabled), and try the connection again?

If you aren't able to sort this out, please find my email address on this page: https://github.com/gitbls/sdm and email me a screen shot of your Mac VPN configuration.

**Comment by gitbls** [comment]
In case it’s not completely clear, here are the sequence of steps to go through to set up your VPN. Each of these steps must be successfully completed. If you’re not sure what to do for any of them, ask!

- Set up port forwarding from your router to your VPN server for ports 500 and 4500 (both UDP)
- Make sure that DNS is properly configured for your VPN server
- Install pistrong with InstallPiStrong from this GitHub
- Run makeMyCA to configure your CA
- Use `sudo pistrong add` to create the certs for your client device
- Install the Certs on your client device
- Test

**Comment by daniel-dlds** [comment]
Hello,
My host public domain name is dlsa.zapto.org. But when I run makeMyCA it prepends the hostname of the pi to the domain. If I say the domain is zapto.org, it thinks the hostname is vpnserver.zapto.org, but is in fact dlsa.zapto.org.
How do I configure the hostname correctly ?

**Reply by gitbls** [reply]
I just went through a makeMyCA on a clean system. The console output contains:
```

External DNS Name or IP Address for your VPN []: dlsa.zapto.org

% 'dlsa.zapto.org' does not appear to be a DNS name known on the external Internet
  The CA will be built with the information you provide but you will
  not be able to connect to your VPN until this is corrected
  Continuing...

** DNS name disambiguation

   Answer Y if 'dlsa' is a hostname and domain name is 'zapto.org'
   Answer N if there is no hostname and domain name is 'dlsa.zapto.org'

Is 'dlsa' a hostname in the DNS domain 'zapto.org' [y/N]? y

% The host name of this system is 'p83' which is different than the VPN DNS name 'dlsa'
  The CA will be built with the information you provided, but you  should consider
  setting your system and VPN host names to be the same to minimize confusion
  Continuing...

** Configuring strongSwan for VPN Server 'p83.zapto.org'
```

The first message indicates that dlsa.zapto.org is not a DNS name known on the internet, presumably because you haven't established it yet? If you think you have it set up already, there's the first thing you need to fix.

The second message (p83 is different than dlsa) tells you what you should do to minimize confusion. Is there some reason that this is not a viable change for you to make?

I went through the rest of the makeMyCA, and the configuration came out with:
```
vpnaddr         dlsa.zapto.org      # This is shown via sudo pistrong config
```
The certificates were also correct, in that the altNames in the certs contained dlsa.zapto.org, which they should. At the same time, the VPN host keys contained
```
  subject:  "C=US, O=p83-strongSwan, CN=p83.zapto.org"
  issuer:   "C=US, O=p83-strongSwan, CN=strongSwan p83 Root CA"
```
I believe that these are correct and should work. However, I don't have any way to easily test that without setting up a full test configuration. I will be willing to do that once you respond to all my previous questions, as well as demonstrating that the configuration generated from makeMyCA is incorrect.

So far, you haven't done that. As I mentioned, the DNS name you want to use is not available on the Internet, and you haven't demonstrated that a client connection is failing because your DNS name is not your hostname.

I look forward to getting complete information from you to help you further.

**Comment by daniel-dlds** [comment]
I changed the hostname in the dynamic dns provider to be vpnserver.zapto.org. Now the mac (that is on the same lan as the strong swan server), goes through the internet using the domain name, and contacts the vpn server. Here is what I get in the server's log :

Mar 30 22:24:07 vpnserver charon: 07[NET] received packet: from 144.64.133.118[500] to 192.168.1.100[500] (604 bytes)
Mar 30 22:24:07 vpnserver charon: 07[ENC] parsed IKE_SA_INIT request 0 [ SA KE No N(REDIR_SUP) N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) ]
Mar 30 22:24:07 vpnserver charon: 07[IKE] no IKE config found for 192.168.1.100...144.64.133.118, sending NO_PROPOSAL_CHOSEN
Mar 30 22:24:07 vpnserver charon: 07[ENC] generating IKE_SA_INIT response 0 [ N(NO_PROP) ]
Mar 30 22:24:07 vpnserver charon: 07[NET] sending packet: from 192.168.1.100[500] to 144.64.133.118[500] (36 bytes)

I installed the pem and pkcs12 certs in the system keychain on the mac.

**Reply by gitbls** [reply]
Did you actually install the certs on the Mac? If so, please email me (per previous instructions) the screenshot I requested demonstrating the VPN configuration on your Mac client.

**Reply by daniel-dlds** [reply]
after the "pistrong add dlsa --device=faisca --pass=XXXXX" the output indicated the path to the zip file. I copied it to the mac, and dragged the pem and the pkcs12 to the system keychain. Below is a screen shot of the vpn connection window on the mac :
<img width="671" alt="Screenshot 2023-03-31 at 18 03 19" src="https://user-images.githubusercontent.com/4472861/229184964-81e2cfbf-d5b7-43bf-ae65-921385a2c1fb.png">
How can I get more information on the client side of what happened ?

**Reply by gitbls** [reply]
A couple of points:

- When you added the certs to the system keychain, did you do this step: "For each of the two just-installed Certs, double-click on them in the Keychain Access.app, expand the Trust section, and choose Always Trust". This MUST be done.
- The Local ID is wrong. You haven't provided me complete information, but unless you changed something from the diagnostics you previously sent, it should be: dlsa-faisca@zapto.org, but you can verify this by looking at the file readme.txt in the zip file. The field is "Local ID/Client ID".

Please make the above change(s) and try it again. 

I have no clue how to get information on the client side. My fingers and eyes don't do Mac. You have a Mac in front of you, so you're the Mac expert here, 

We can look at it from the server side.

If it doesn't work, I need to see your system log: `sudo journalctl -b | grep charon`, and also please provide the output from `sudo pistrong list --full`

One other thing: if it still doesn't work, please provide a screen shot of the Mac VPN connection showing the "Authentication settings" dialog.

**Reply by daniel-dlds** [reply]
I managed to look at the logs on the mac. The screenshot from Console app with the relevant lines are below. Seems to me that the error is the same as on the server, but I have no clue
<img width="1422" alt="Screenshot 2023-04-01 at 18 09 58" src="https://user-images.githubusercontent.com/4472861/229305020-efc878e5-41a3-4ed2-9bd0-53e6db003bfc.png">

**Reply by gitbls** [reply]
I'm sorry, I need ALL the information I requested above. It's clearly a configuration error of some sort, but without the information I requested, I doubt that I'll be able to help you. Please provide ALL the requested information.

**Reply by daniel-dlds** [reply]
I am sending you the result of pscollect since the last time I regenerated the certificates. Please tell me what more do you need.
[pistrong-diagnostics.zip](https://github.com/gitbls/pistrong/files/11133053/pistrong-diagnostics.zip)
From what I see in the vpnserver's logs, the message is NO PROPOSAL CHOSEN. Is this related with the available cyphers on each side not being compatible ?

**Reply by gitbls** [reply]
I will have a look at the diagnostics, but won't reply until you provide the rest of the information that I requested, because the answer will be in the stuff you haven't provided yet. Specifically:

- The system log from vpnserver showing a failed connect from your Mac. Telling me NO PROPOSAL CHOSEN is ony part of the situation. I need ALL the log entries from the connect
- The answer to these two questions:
    - When you added the certs to the system keychain, did you do this step: "For each of the two just-installed Certs, double-click on them in the Keychain Access.app, expand the Trust section, and choose Always Trust". This MUST be done.
    - The Local ID is wrong. You haven't provided me complete information, but unless you changed something from the diagnostics you previously sent, it should be: [dlsa-faisca@zapto.org](mailto:dlsa-faisca@zapto.org), but you can verify this by looking at the file readme.txt in the zip file. The field is "Local ID/Client ID".
- A current screenshot of the Mac VPN configuration
- A current screenshot of the Mac's "Authentication Settings" dialog (hit the _Authentication Settings_ button in the VPN configuration dialog
- A copy of the readme.txt from the pistrong-generated Certs zip file for your Mac client.

Once I have all the above information, I'll be happy to look into this further. Please provide ALL the above information so we can figure out what's wrong with your system and get you going.

Without all this information, I'd only be guessing which will do neither of us any good.

**Reply by daniel-dlds** [reply]
1st item, the log entries from the stronsean server : 
[strongswan.log](https://github.com/gitbls/pistrong/files/11133458/strongswan.log)
2nd item
I set both certificates to Trusted
I set the local ID to the one in the readme.txt that is inside the zip file generated in the add command
Conection Details:
   Server:                vpnserver.zapto.org
   Remote ID/Server ID:   ios.zapto.org
   Local ID/Client ID:    dlsa-faisca-vpnserver@vpnserver.zapto.org
   Cert Name:             dlsa-faisca-vpnserver@vpnserver.zapto.org
Current screen shot of mac vpn
<img width="668" alt="Screenshot 2023-04-02 at 23 10 29" src="https://user-images.githubusercontent.com/4472861/229382142-b0a3fd8e-93e5-4e40-9be6-acd0c6c90f68.png">
current screenshot of the macs auth settings : 
<img width="383" alt="Screenshot 2023-04-02 at 23 13 11" src="https://user-images.githubusercontent.com/4472861/229382159-46c78f91-4ca7-473d-8c2b-4716fb5bcea7.png">
certificate selected in auth settings : 
<img width="536" alt="Screenshot 2023-04-02 at 23 13 22" src="https://user-images.githubusercontent.com/4472861/229382189-65f0fa58-4c7d-4754-9f31-aa38caa8e46e.png">
copy of the readme file in the generated zip file : 
provided above.
Thanks for all your help

**Reply by gitbls** [reply]
Thanks! This helps a lot and rules out many potential issues. It appears that your configuration on the Mac is now correct.

I dont' see any connects in the strongswan.log you sent me that appear to be from a properly-configured client. Soo....

What do you get when you run the following command on your Mac before trying to connect to the VPN?
```
ping vpnserver.zapto.org
```
And, how is your Mac connected to the internet?

If it's on the same LAN as vpnserver that will only work with a very few routers that support hairpinning. You can test your VPN on the same LAN if and only if the DNS name vpnserver.zapto.org resolves on the Mac to the LAN address of your vpn server.

Alternatively, if you can tether via your phone you'll be outside the LAN and that will work.

One other thing: Have you rebooted the vpnserver since running makeMyCA? If you have not, please do that.

**Reply by daniel-dlds** [reply]
I was connecting from inside the lan using the dynamic DNS from the internet. I tried connecting form outside the LAN and got the following in the logs : 
[mac_connect_log.log](https://github.com/gitbls/pistrong/files/11141091/mac_connect_log.log)
[strongswan.log](https://github.com/gitbls/pistrong/files/11141092/strongswan.log)

**Reply by daniel-dlds** [reply]
its the NO_PROPOSAL_CHOSEN message. Can it be that the strongswan server cannot agree with the mac the encryption algorithm to use ?

**Reply by gitbls** [reply]
You didn't answer my question as to whether you can ping vpnserver.zapto.org from your Mac. Please address that question. It looks to me in the strongswan log like your Mac has not even tried to connect to the server.

I misse dthe part abou trying from outside the LAN. What was your Mac's IP address at that time? Without knowing that I can't tell WTF is going on.

In other words, I can't tell whether the no proposal chosen message was for your Mac, or someone else trying to break in (there are quite a few attempts from other random ip addresses). 

When your Mac connects, I expect to see some lines similar to these in the vpn server log:
```
Mar 22 14:23:14 pvn charon-systemd[665]: 166.199.5.115 is initiating an IKE_SA
Mar 22 14:23:14 pvn charon-systemd[665]: selected proposal: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048
Mar 22 14:23:14 pvn charon-systemd[665]: looking for peer configs matching 192.168.92.5[ios.xyzzy.com]...166.199.5.115[bls-biphone-pvn@myvpn.net]
Mar 22 14:23:14 pvn charon-systemd[665]: selected peer config 'conn-ios'
Mar 22 14:23:14 pvn charon-systemd[665]: negotiated TLS 1.2 using suite TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
```

**Comment by gitbls** [comment]
Please do the following test with your Mac not on your local LAN:

- In a command window, type: curl -s 'dynupdate.no-ip.com/ip.php'
- In that same command window type: ping vpnserver.zapto.org
- Provide the output from those two commands
- Try to connect to the VPN
- Provide the output from the Mac and vpn server as you just did.

The above test will tell us if your Mac is actually trying to connect to the VPN server (the logs seem to indicate that it's not, but without additional information (such as the 2 commands above) it's difficult to say with certainty.

**Comment by gitbls** [comment]
Hi, what version of MacOS do you have? If it's MacOS Ventura, then an edit is needed to the VPN configuration. 
In the file /etc/swanctl/conf.d/pistrong-CAServerConnection.conf, change these two lines:
```
    conn-ios : conn-defaults, remote-defaults {
        proposals = aes256-sha256-modp2048, aes256-sha256-modp1024,aes256-sha1-modp1024
```
to be
```
    conn-ios : conn-defaults, remote-defaults {
        proposals = aes256-sha256-modp2048, aes256-sha256-modp1024,aes256-sha1-modp1024,aes256-sha256-prfsha256-modp2048
```
In brief: add aes256-sha256-prfsha256-modp2048 to the end of the "proposals" line, with a comma in front of it 

Then `sudo systemctl restart strongswan` and try to connect again from your Mac.

Please let me know the result.

If this fixes it, you were correct that it was a change in the encryption algorithm requirements in MacOS.

**Reply by daniel-dlds** [reply]
Hi, I am on Monterey, with the latest updates. I added that proposal to the list and that did the trick, I can now connect to the VPN inside or outside the vpnserver's LAN.
Thanks a lot for your help.

**Reply by gitbls** [reply]
Great! Thanks for sticking with me on solving this and enjoy your VPN!

I'll check in an updated makeMyCA this week with this update. EDIT: This has been checked in.
